dev/NavigationView/NavigationView.cpp

```diff
@@ -379,10 +427,354 @@ void NavigationView::OnApplyTemplate()
+
+void NavigationView::UpdateItemsRepeaterItemsSource(const winrt::ItemsRepeater& ir,
+    const winrt::IInspectable& itemsSource)
+{
+    if (ir)
+    {
+        auto oldItemsSource = ir.ItemsSource();
+        if (oldItemsSource != itemsSource)
+        {
+            ir.ItemsSource(itemsSource);
+        }
+    }
+}
+
+void NavigationView::OnNavigationViewItemIsSelectedPropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    auto nvib = sender.try_as<winrt::NavigationViewItemBase>();
+    if (nvib)
+    {
+        // Check whether the container that triggered this call back is the selected container
+        bool newItemIsSelectedItem = IsContainerTheSelectedItemInTheSelectionModel(nvib);
+        bool newIsSelected = nvib.IsSelected();
+
+        if (newIsSelected && !newItemIsSelectedItem)
+        {
+            winrt::IndexPath ip = GetIndexPathForContainer(nvib);
+            m_selectionModel.SelectAt(ip);
+        }
+        else if (!newIsSelected && newItemIsSelectedItem)
+        {
+            auto indexPath = GetIndexPathForContainer(nvib);
+            auto indexPathFromModel = m_selectionModel.SelectedIndex();
+
+            if (indexPath && indexPathFromModel && indexPath.GetAt(0) == indexPathFromModel.GetAt(0))
+            {
+                m_selectionModel.DeselectAt(indexPath);
+            }
+        }
+    }
+}
+
+void NavigationView::RaiseItemInvokedForNavigationViewItem(const winrt::NavigationViewItem& nvi)
+{
+    winrt::IInspectable nextItem = nullptr;
+    auto prevItem = SelectedItem();
+    auto parentIR = GetParentItemsRepeaterForContainer(nvi);
+
+    bool isInOverflow = parentIR.Name() == c_overflowRepeater;
+    bool itemSelectsOnInvoked = nvi.SelectsOnInvoked();
+
+    auto itemIndex = parentIR.GetElementIndex(nvi);
+    auto itemsSourceView = parentIR.ItemsSourceView();
+    if (itemsSourceView)
+    {
+        auto inspectingDataSource = static_cast<InspectingDataSource*>(winrt::get_self<ItemsSourceView>(itemsSourceView));
+        nextItem = inspectingDataSource->GetAt(itemIndex);
+    }
+
+    // Other transition other than default only apply to topnav
+    // when clicking overflow on topnav, transition is from bottom
+    // otherwise if prevItem is on left side of nextActualItem, transition is from left
+    //           if prevItem is on right side of nextActualItem, transition is from right
+    // click on Settings item is considered Default
+    NavigationRecommendedTransitionDirection recommendedDirection = NavigationRecommendedTransitionDirection::Default;
+    if (IsTopNavigationView() && itemSelectsOnInvoked)
+    {
+        if (isInOverflow)
+        {
+            recommendedDirection = NavigationRecommendedTransitionDirection::FromOverflow;
+        }
+        else if (prevItem && nvi)
+        {
+            recommendedDirection = GetRecommendedTransitionDirection(NavigationViewItemBaseOrSettingsContentFromData(prevItem), nvi);
+        }
+    }
```

Comment:
Try to avoid scratch variables, per ES.28 this should probably be defined using a lambda [https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es21-dont-introduce-a-variable-or-constant-before-you-need-to-use-it](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es21-dont-introduce-a-variable-or-constant-before-you-need-to-use-it)